packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -38,27 +98,193 @@ contract FundingRateStore is FundingRateStoreInterface, Testable {
+     * @dev Caller must approve this this contract to spend `finalFeeBond` amount of collateral, which they can
+     * receive back if their dispute is successful.
+     */
+    function dispute(bytes32 identifier) external {
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(identifier);
+        ProposalState proposalState = _getProposalState(fundingRateRecord.proposal);
+
+        require(proposalState == ProposalState.Pending, "Existing proposal not pending.");
+
+        // TODO: Bond logic.
+
+        // Pull final fee bond from disputer and pay store.
+        _payFinalFees(msg.sender, fundingRateRecord.proposal.finalFee);
+
+        // Send price request
+        _requestOraclePrice(identifier, fundingRateRecord.proposal.time);
+
+        emit DisputedRate(
+            identifier,
+            fundingRateRecord.proposal.rate.rawValue,
+            fundingRateRecord.proposal.time,
+            fundingRateRecord.proposal.proposer,
+            msg.sender
+        );
+
+        // Delete pending proposal and copy into dispute records.
+        fundingRateRecord.proposal.disputer = msg.sender;
+        fundingRateDisputes[identifier][fundingRateRecord.proposal.time] = fundingRateRecord;
+        delete fundingRateRecords[identifier].proposal;
+    }
+
+    /**
+     * @notice Settle a disputed funding rate. The winner of the dispute, either the disputer or the proposer,
+     * will receive a reward plus their final fee bond. This method will also overwrite the current funding rate
+     * with the resolved funding rate returned by the Oracle. Pending funding rates are unaffected by this method.
+     * @dev This will revert if there is no price available for the disputed funding rate.
+     */
+    function settleDispute(bytes32 identifier, uint256 proposalTime) external {
+        FundingRateRecord storage fundingRateDispute = _getFundingRateDispute(identifier, proposalTime);
+
+        // Get the returned funding rate from the oracle. If this has not yet resolved will revert.
+        // If the fundingRateDispute struct has been deleted, then this call will also fail because the proposal
+        // time will be 0.
+        FixedPoint.Signed memory settlementRate = _getOraclePrice(identifier, fundingRateDispute.proposal.time);
+
+        // Dispute was successful if settled rate is different from proposed rate.
+        bool disputeSucceeded = !settlementRate.isEqual(fundingRateDispute.proposal.rate);
+        address proposer = disputeSucceeded
+            ? fundingRateDispute.proposal.disputer
+            : fundingRateDispute.proposal.proposer;
+
+        if (disputeSucceeded) {
+            // If dispute succeeds:
+            // - pay disputer the dispute reward + final fee rebate
+
+            collateralCurrency.safeTransfer(
+                fundingRateDispute.proposal.disputer,
+                fundingRateDispute.proposal.finalFee.rawValue
```

Comment:
how is this sending the `finalFee + dispute reward`? I'm guessing this will come in later within the `fundingRateRecords` structure.

Reply:
There is no dispute reward yet as you guess. We would add it in here.